storage/repository/repository.go

```diff
@@ -0,0 +1,198 @@
+package repository
+
+import (
+	"fmt"
+
+	"fyne.io/fyne"
+)
+
+// Repository represents a storage repository, which is a set of methods which
+// implement specific functions on a URI. Repositories are registered to handle
+// specific URI schemes, and the higher-level functions that operate on URIs
+// internally look up an appropriate method from the relevant Repository.
+//
+// The repository interface includes only methods which must be implemented at
+// a minimum. Without implementing all of the methods in this interface, a URI
+// would not be usable in a useful way. Additional functionality can be exposed
+// by using interfaces which extend Repository.
+//
+// Repositories are registered to handle a specific URI scheme (or schemes)
+// using the Register() method. When a higher-level URI function such as
+// storage.Copy() is called, the storage package will internally look up
+// the repository associated with the scheme of the URI, then it will use
+// a type assertion to check if the repository implements CopyableRepository.
+// If so, the Copy() function will be run from the repository, otherwise
+// storage.Copy() will return NotSupportedError. This works similarly for
+// all other methods in repository-related interfaces.
+//
+// Note that a repository can be registered for multiple URI schemes. In such
+// cases, the repository must internally select and implement the correct
+// behavior for each URI scheme.
+//
+// NOTE: most developers who use Fyne should *not* generally attempt to
+// call repository methods directly. You should use the methods in the storage
+// package, which will automatically detect the scheme of a URI and call into
+// the appropriate repository.
+//
+// Since: 2.0.0
+type Repository interface {
+
+	// Exists will be used to implement calls to storage.Exists() for the
+	// registered scheme of this repository.
+	//
+	// Since 2.0.0
+	Exists(u fyne.URI) (bool, error)
+
+	// ReaderFrom will be used to implement calls to storage.ReaderFrom()
+	// for the registered scheme of this repository.
+	//
+	// Since 2.0.0
+	ReaderFrom(u fyne.URI) (fyne.URIReadCloser, error)
+
+	// CanRead will be used to implement calls to storage.CanRead() for the
+	// registered scheme of this repository.
+	//
+	// Since 2.0.0
+	CanRead(u fyne.URI) (bool, error)
+
+	// Destroy is called when the repository is un-registered from a given
+	// URI scheme.
+	//
+	// Since 2.0.0
+	Destroy()
```

Comment:
Should this have a parameter of the scheme to which the Repository is no longer registered?

Reply:
Also a good idea; that would matter for some repositories.

Reply:
Done in 188d6764d48a731559a84f349ea959e35798e0a8